hazelcast/src/main/java/com/hazelcast/cache/impl/CacheSplitBrainHandlerService.java

```diff
@@ -144,8 +155,79 @@ private boolean isLocalPartition(int partitionId) {
 
         @Override
         public void run() {
-            final Semaphore semaphore = new Semaphore(0);
+            int recordCount = 0;
+            for (Map.Entry<String, Map<Data, CacheRecord>> recordMapEntry : recordMap.entrySet()) {
+                String cacheName = recordMapEntry.getKey();
+                Map<Data, CacheRecord> records = recordMapEntry.getValue();
+                Object mergePolicy = getCacheMergePolicy(cacheName);
+                if (mergePolicy instanceof SplitBrainMergePolicy) {
+                    // we cannot merge into a 3.9 cluster, since not all members may understand the MergeOperationFactory
+                    if (nodeEngine.getClusterService().getClusterVersion().isLessThan(Versions.V3_10)) {
+                        logger.info("Cannot merge cache '" + cacheName
+                                + "' with merge policy '" + mergePolicy.getClass().getName()
+                                + "' until cluster is running version " + Versions.V3_10);
+                        continue;
+                    }
+                    recordCount += handleMerge(cacheName, records, (SplitBrainMergePolicy) mergePolicy, DEFAULT_BATCH_SIZE);
+                } else {
+                    recordCount += handleMerge(cacheName, records, (CacheMergePolicy) mergePolicy);
+                }
+            }
+            recordMap.clear();
+
+            try {
+                semaphore.tryAcquire(recordCount, recordCount * TIMEOUT_FACTOR, TimeUnit.MILLISECONDS);
+            } catch (InterruptedException e) {
+                logger.finest("Interrupted while waiting merge operation...");
+            }
+        }
+
+        private int handleMerge(String name, Map<Data, CacheRecord> recordMap, SplitBrainMergePolicy mergePolicy, int batchSize) {
```

Comment:
is it possible to break down this method into smaller methods to increase readability?

Reply:
The main issue I see is that a lot of local variables are defined and used. This is hard to split up. For example this block would be a nice candidate:
```java
            MutableLong[] counterPerMember = new MutableLong[partitionCount];
            Address[] addresses = new Address[partitionCount];
            for (Map.Entry<Address, List<Integer>> addressListEntry : memberPartitionsMap.entrySet()) {
                MutableLong counter = new MutableLong();
                Address address = addressListEntry.getKey();
                for (int partitionId : addressListEntry.getValue()) {
                    counterPerMember[partitionId] = counter;
                    addresses[partitionId] = address;
                }
            }
```
But it defines two arrays, so you cannot just do a nice single line method pull out like `MutableLong[] counterPerMember = createCounter(partitionCount);`.

Reply:
ok i see. what about adding some explanatory comments for the goal of each block?

Reply:
I've added some comments (will do the same for the map and replicated map PR, which have the same logic).